Added rust-analyzer instructions for Helix (#3936)
rust-lang/miri: CONTRIBUTING.md

## Configuring `rust-analyzer`

To configure `rust-analyzer` and the IDE for working on Miri, use one of the provided
configuration files according to the instructions below.

### Visual Studio Code

> Copy [`src/etc/rust_analyzer_settings.json`] to `.vscode/settings.json` in the project root directory.

Reviewer: The paths are still pointing to `src/`

### Helix

Copy [`src/etc/rust_analyzer_helix.toml`] to `.helix/languages.toml` in the project root directory.

> Since working on Miri requires a custom toolchain consider running `./miri toolchain` with the
> `-c rust-analyzer` flag, or set the `RUSTUP_TOOLCHAIN` environment variable according to

Reviewer: ```suggestion
Since working on Miri requires a custom toolchain, and Helix requires the language server
to be installed with the toolchain, you have to run `./miri toolchain -c rust-analyzer`
when installing the Miri toolchain. Alternatively, set the `RUSTUP_TOOLCHAIN` environment variable according to
```

> [the documentation](https://rust-analyzer.github.io/manual.html#toolchain).

Author: Added a note how to properly configure `rust-analyzer` for working on Miri. Another option would be following up the referenced documentation and setting the environment variable in the provided config, but I'm not sure about that. It seems too implicit, and in theory `rust-analyzer` might not be able to support everything.

> #### Hint
>
> To keep the `rust-analyzer` configuration up-to-date, make a symbolic link to the provided file.

> > #### Note

Reviewer: This now looks like the note is specific for Helix, which is not the case.

Any suggestion for how to avoid that confusion?

Author: I haven't decided how it is better yet. Moving it before the IDE paragraphs makes it too bold and the follow up seems disconnected. I will think about it.

Author: Seems better now. Are you OK with it?

Reviewer: I would suggest to move the hint (about using a symlink) to the first section, the one about VSCode. People reading to-down will hopefully notice it then even if they are using a different IDE.

The note about "If you are building Miri with a locally built rustc" could move into the "Advanced configuration" section.

Author: > People reading to-down will hopefully notice it then even if they are using a different IDE.

Maybe above the `### Visual Studio Code`? Then it wouldn't look like it applies only to VS Code, but also for Helix.

Reviewer: Then it is confusing because there haven't been any "files" yet that this comment refers to.

Author: Make a symbolic link to the provided files below? Still the same place, but it point that the files are inside the IDE and Helix sections.

If you're not satisfied with that idea, I will just follow your suggestion to get it merged.

Reviewer: I don't understand the suggestion. But a forward reference will just be confusing I think, this hint can't come before the first concrete IDE configuration section.

>

> > If you are [building Miri with a locally built rustc][], set

Reviewer: This link does not work, it seems. Probably pre-existing, but since you are touching this anyway...
```suggestion
> If you are building Miri with a locally built rustc, set
```

> `rust-analyzer.rustcSource` to the relative path from your Miri clone to the
> root `Cargo.toml` of the locally built rustc. For example, the path might look

> > like `../rust/Cargo.toml`. In addition to that replace `clippy` by `check` command

Reviewer: ```suggestion
> like `../rust/Cargo.toml`. In addition to that, replace `clippy` by `check`
```

## Configuring `rust-analyzer`

To configure `rust-analyzer` and the IDE for working on Miri, use one of the provided
configuration files according to the instructions below.

> #### Note
>
> If you are building Miri with a locally built rustc, set
> `rust-analyzer.rustcSource` to the relative path from your Miri clone to the
> root `Cargo.toml` of the locally built rustc. For example, the path might look
> like `../rust/Cargo.toml`. In addition to that, replace `clippy` by `check`
> in the `rust-analyzer.check.overrideCommand` setting.

### Visual Studio Code

> Copy [`etc/rust_analyzer_settings.json`] to `.vscode/settings.json` in the project root directory.

Reviewer: What is the point of the brackets around the filename?

These do not become links like this.

Author: I thought that it was a style in [rustc-dev-guide](https://rustc-dev-guide.rust-lang.org/building/suggested.html#helix), but it seems that there are just broken links.

### Visual Studio Code

Copy [`etc/rust_analyzer_vscode.json`] to `.vscode/settings.json` in the project root directory.

[`etc/rust_analyzer_vscode.json`]: https://github.com/rust-lang/miri/blob/master/etc/rust_analyzer_vscode.json

### Helix

Copy [`etc/rust_analyzer_helix.toml`] to `.helix/languages.toml` in the project root directory.

Since working on Miri requires a custom toolchain, and Helix requires the language server
to be installed with the toolchain, you have to run `./miri toolchain -c rust-analyzer`
when installing the Miri toolchain. Alternatively, set the `RUSTUP_TOOLCHAIN` environment variable according to
[the documentation](https://rust-analyzer.github.io/manual.html#toolchain).

[`etc/rust_analyzer_helix.toml`]: https://github.com/rust-lang/miri/blob/master/etc/rust_analyzer_helix.toml

### Advanced configuration

> #### Note
>
> If you are building Miri with a locally built rustc, set
> `rust-analyzer.rustcSource` to the relative path from your Miri clone to the
> root `Cargo.toml` of the locally built rustc. For example, the path might look
> like `../rust/Cargo.toml`. In addition to that, replace `clippy` by `check`

> > in the `rust-analyzer.check.overrideCommand` setting.

Reviewer: It's odd to make this a "note" but the next thing here not. Please just make them two paragraphs.

Author: It was originally a note, I haven't touched that at all, just moved as it was asked (:

Reviewer: Yeah fair. :) But previously there was only a single IDE so we didn't need any of this structure. If you want I can push push some edits to your PR, that's probably a more effective way to do the final fine-tuning.
